File: polyglot-kotlin/src/main/kotlin/org/sonatype/maven/polyglot/kotlin/dsl/Dependency.kt
package org.sonatype.maven.polyglot.kotlin.dsl

@PomDsl
class Dependency : org.apache.maven.model.Dependency(), Cloneable {

  @PomDsl
  fun groupId(groupId: String): Dependency {
    this.groupId = groupId
    return this
  }

  @PomDsl
  fun artifactId(artifactId: String): Dependency {
    this.artifactId = artifactId
    return this
  }

  @PomDsl
  fun version(version: String): Dependency {
    this.version = version
    return this
  }

  @PomDsl
  fun type(type: String): Dependency {
    this.type = type
    return this
  }

  @PomDsl
  fun classifier(classifier: String): Dependency {
    this.classifier = classifier
    return this
  }

  @PomDsl
  fun scope(scope: String): Dependency {
    this.scope = scope
    return this
  }

  @PomDsl
  fun systemPath(systemPath: String): Dependency {
    this.systemPath = systemPath
    return this
  }

  @PomDsl
  fun optional(optional: Boolean = true): Dependency {
    this.isOptional = optional
    return this
  }

  /** Provides a callback for defining dependency exclusions. */
  @PomDsl
  fun exclusions(block: ExclusionList.(ExclusionList) -> Unit): Dependency {
    val exclusionList = ExclusionList()
    block(exclusionList, exclusionList)
    this.exclusions = exclusionList
    return this
  }

  /**
   * Excludes the artifacts having the specified `groupId:artifactId`. The wildcard `*` character
   * can be used to exclude matching artifacts.
   */
  @PomDsl
  fun exclusions(vararg ga: String): Dependency {
    this.exclusions =
        ga.map { splitCoordinates(it) }
            .map { (groupId, artifactId) ->
              Exclusion().apply {
                this.groupId = groupId
                this.artifactId = artifactId
              }
            }
    return this
  }

  override fun clone(): org.apache.maven.model.Dependency {
    return super<org.apache.maven.model.Dependency>.clone()
  }
}

File: polyglot-scala/src/main/scala/org/sonatype/maven/polyglot/scala/model/Dependency.scala
/**
 * Copyright (c) 2012 to original author or authors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */
package org.sonatype.maven.polyglot.scala.model

import scala.collection.immutable

class Dependency(
    val gav: Gav,
    val `type`: String,
    val classifier: Option[String],
    val scope: Option[String],
    val systemPath: Option[String],
    val exclusions: immutable.Seq[GroupArtifactId],
    val optional: Boolean
) {

  /**
   * Returns a derived dependency with the given new properties.
   */
  def copy(
      gav: Gav = gav,
      `type`: String = `type`,
      classifier: Option[String] = this.classifier,
      scope: Option[String] = this.scope,
      systemPath: Option[String] = this.systemPath,
      exclusions: immutable.Seq[GroupArtifactId] = exclusions,
      optional: Boolean = optional
  ): Dependency =
    new Dependency(gav, `type`, classifier, scope, systemPath, exclusions, optional)

  /**
   * Returns a derived dependencies with the given classifier.
   */
  def classifier(classifier: String): Dependency = copy(classifier = Option(classifier))

  /**
   * Returns a derived dependency without it's transitive dependencies.
   * This is internally done by setting an universal exclusion (`"*" % "*"`).
   */
  def intransitive: Dependency = copy(exclusions = immutable.Seq("*" % "*"))

  /**
   * Returns a derived dependency with the given scope.
   */
  def %(scope: String): Dependency = copy(scope = Option(scope).filter(s => !s.trim().isEmpty()))

}

object Dependency {
  def apply(
      gav: Gav,
      `type`: String = "jar",
      classifier: String = null,
      scope: String = null,
      systemPath: String = null,
      exclusions: immutable.Seq[GroupArtifactId] = immutable.Seq.empty,
      optional: Boolean = false
  ) =
    new Dependency(
      gav,
      `type`,
      Option(classifier),
      Option(scope),
      Option(systemPath),
      exclusions,
      optional
    )
}

import org.sonatype.maven.polyglot.scala.ScalaPrettyPrinter._

class PrettiedDependency(d: Dependency) {
  def asDoc: Doc = {
    val typeAssigned = d.`type` != "jar"
    val classifierAssigned = d.classifier.isDefined
    val systemPathAssigned = d.systemPath.isDefined
    val exclusionsAssigned = d.exclusions.nonEmpty
    val optionalAssigned = d.optional
    if (
      typeAssigned || classifierAssigned || systemPathAssigned || exclusionsAssigned || optionalAssigned
    ) {
      val args = scala.collection.mutable.ListBuffer(d.gav.asDoc)
      if (typeAssigned) args += assignString("`type`", d.`type`)
      d.classifier.foreach(args += assignString("classifier", _))
      d.scope.foreach(args += assignString("scope", _))
      d.systemPath.foreach(args += assignString("systemPath", _))
      if (exclusionsAssigned) args += assign("exclusions", seq(d.exclusions.map(_.asDoc)))
      if (optionalAssigned) args += assign("optional", d.optional.toString)
      `object`("Dependency", args.toList)
    } else {
      val gav = d.gav.asDoc
      val version = if (d.gav.version.isEmpty) space <> percent <+> dquote <> dquote else emptyDoc
      gav <> version <> d.scope.map(emptyDoc <+> percent <+> dquotes(_)).getOrElse(emptyDoc)
    }
  }
}

import org.sonatype.maven.polyglot.scala.MavenConverters._
import scala.jdk.CollectionConverters._
import org.apache.maven.model.{Dependency => MavenDependency, Exclusion => MavenExclusion}

class ConvertibleMavenDependency(md: MavenDependency) {
  def asScala: Dependency = {
    Dependency(
      (md.getGroupId, md.getArtifactId, md.getVersion).asScala,
      md.getType,
      md.getClassifier,
      md.getScope,
      md.getSystemPath,
      md.getExclusions.asScala.map(e => (e.getGroupId, e.getArtifactId).asScala).toList,
      md.isOptional
    )
  }
}

class ConvertibleScalaDependency(d: Dependency) {
  def asJava: MavenDependency = {
    val md = new MavenDependency
    md.setArtifactId(d.gav.artifactId)
    md.setClassifier(d.classifier.orNull)
    md.setExclusions(d.exclusions.map({
      gav =>
        val e = new MavenExclusion
        e.setArtifactId(gav.artifactId)
        e.setGroupId(gav.groupId.orNull)
        e
    }).asJava)
    md.setGroupId(d.gav.groupId.orNull)
    md.setOptional(d.optional)
    md.setScope(d.scope.orNull)
    md.setSystemPath(d.systemPath.orNull)
    md.setType(d.`type`)
    md.setVersion(d.gav.version.orNull)
    md
  }
}
